File: frontend/src/api/packets.ts
import {
    Vec,
    bool,
    ByteArray,
    MapType,
    PacketDefinition,
    Str,
    StructVec,
    u32,
    u8,
    VarInt
} from "wsbps-js"

// The different possible values for player data packet modes
export enum PlayerDataMode {ADD, REMOVE, SELF}

// An enum containing different states the client can request
// from the server
export enum States {DISCONNECT, START, SKIP}

// SERVER PACKETS
export const DisconnectPacket = new PacketDefinition(0x00, {reason: Str}, ['reason']);
export const ErrorPacket = new PacketDefinition(0x01, {cause: Str}, ['cause']);
export const JoinGamePacket = new PacketDefinition(0x02, {id: Str, owner: bool, title: Str}, ['id', 'owner', 'title']);
export const NameTakenResultPacket = new PacketDefinition(0x03, {result: bool}, ['result']);
export const GameStatePacket = new PacketDefinition(0x04, {state: u8}, ['state']);
export const PlayerDataPacket = new PacketDefinition(0x05, {id: Str, name: Str, mode: u8}, ['id', 'name', 'mode']);
export const TimeSyncPacket = new PacketDefinition(0x06, {total: VarInt, remaining: VarInt}, ['total', 'remaining']);
export const QuestionPacket = new PacketDefinition(0x07, {
    imageType: Str,
    image: ByteArray,
    question: Str,
    answers: Vec(Str),
}, ['image', 'question', 'answers']);
export const AnswerResultPacket = new PacketDefinition(0x08, {result: bool}, ['result']);
export const ScoresPacket = new PacketDefinition(0x09, {scores: MapType(Str, u32)}, ['scores']);

// CLIENT PACKETS
export const CreateGamePacket = new PacketDefinition(0x00, {
    title: Str,
    questions: StructVec({
        imageType: Str,
        image: ByteArray,
        question: Str,
        answers: Vec(Str),
        values: Vec(u8)
    }, ['imageType', 'image', 'question', 'answers', 'values'])
}, ['title', 'questions']);
export const CheckNameTakenPacket = new PacketDefinition(0x01, {id: Str, name: Str}, ['id', 'name']);
export const RequestGameStatePacket = new PacketDefinition(0x02, {id: Str}, ['id']);
export const RequestJoinPacket = new PacketDefinition(0x03, {id: Str, name: Str}, ['id', 'name']);
export const StateChangePacket = new PacketDefinition(0x04, {state: u8}, ['state']);
export const AnswerPacket = new PacketDefinition(0x05, {id: u8}, ['id']);
export const KickPacket = new PacketDefinition(0x06, {id: Str}, ['id']);

File: backend/src/packets.rs
use std::collections::HashMap;
use wsbps::{packet_data, packets, VarInt};

packet_data! {

    enum GameState (->) (u8) {
        Waiting: 0,
        Starting: 1,
        Started: 2,
        Stopped: 3,
        DoesNotExist: 4
    }

    enum PlayerDataMode (->) (u8) {
        Add: 0,
        Remove: 1,
        Me: 2
    }

    enum StateChange (<-) (u8) {
        Disconnect: 0,
        Start: 1,
        Skip: 2
    }

    struct QuestionData (<-) {
        image_type: String,
        image: Vec<u8>,
        question: String,
        answers: Vec<String>,
        values: Vec<u8>
    }
}

pub type ScoresMap = HashMap<String, u32>;

packets! {
    ServerPackets (->) {
        SDisconnect (0x00) { reason: String }
        SError (0x01) { cause: String }
        SJoinedGame (0x02) { id: String, owner: bool, title: String}
        SNameTakenResult (0x03) { result: bool }
        SGameState (0x04) { state: GameState }
        SPlayerData (0x05) { id: String, name: String, mode: PlayerDataMode }
        STimeSync (0x06) { total: VarInt, remaining: VarInt}
        SQuestion (0x07) { image: Vec<u8>, question: String, answers: Vec<String> }
        SAnswerResult (0x08) { result: bool }
        SScores (0x09) { scores: ScoresMap }
    }

    ClientPackets (<-) {
        CCreateGame (0x00) { title: String, questions: Vec<QuestionData> }
        CCheckNameTaken (0x01) { id: String, name: String}
        CRequestGameState (0x02) { id: String }
        CRequestJoin (0x03) { id: String, name: String }
        CStateChange (0x04) { state: StateChange }
        CAnswer (0x05) { id: u8 }
        CKick (0x06) { id: String }
    }
}
